javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,951 @@
 package java.lang
 
 import java.io.Serializable
+import java.util.Comparator
+import java.util.Formatter
+import java.util.Locale
+import java.util.regex.Pattern
+import java.nio.ByteBuffer
+import java.nio.CharBuffer
+import java.nio.charset.Charset
+import java.nio.charset.IllegalCharsetNameException
+import java.nio.charset.UnsupportedCharsetException
+import java.security.AccessController
+import scala.util.Try
 
-class _String private() extends CharSequence with Serializable with Comparable[_String] {
-  def this(value: Array[scala.Char], offset: scala.Int, count: scala.Int) = this()
-
-  def length(): scala.Int = ???
-  def charAt(index: scala.Int): Char = ???
-  def compareTo(another_String: _String): scala.Int = ???
-  def subSequence(beginIndex: scala.Int, endIndex: scala.Int): CharSequence = ???
-  def concat(str: _String): _String = ???
-  def indexOf(ch: scala.Int): scala.Int = ???
-  def indexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+class _String extends Serializable with Comparable[_String] with CharSequence {
+  private var value = new Array[Char](0)
+  private var offset = 0
+  private var count = 0
+  private override var hashCode: Int = _
+
+  private def this(s: _String, ch: Char) {
+    this()
+    offset = 0
+    value = Array.ofDim[Char](s.count + 1)
+    count = s.count + 1
+    System.arraycopy(s.value, s.offset, value, 0, s.count)
+    value(s.count) = ch
+  }
+
+  def this(data: Array[scala.Byte], start: Int, length: Int) {
+    this()
+    if (start >= 0 && 0 <= length && length <= data.length - start) {
+      offset = 0
+      val charset = defaultCharset()
+      val cb = charset.decode(ByteBuffer.wrap(data, start, length))
+      val result: Int = cb.length
+
+      if (result > 0) {
+        value = cb.array()
+        count = result
+      } else {
+        count = 0
+        value = Array.ofDim[Char](0)
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  @Deprecated
+  def this(data: Array[scala.Byte],
+           high: Int,
+           start: Int,
+           length: Int) {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = Array.ofDim[Char](length)
+      count = length
+      high <<= 8
+      for (i <- 0 until count) {
```

Comment:
Please avoid Scala for loops for now and use while loops instead. We're currently not quite there optimisation-wise to remove the overhead they have.